packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1473,6 +1485,137 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
+        if (possibleExistingEdge != null && (possibleExistingEdge == existingSelectionStart && isEnd || possibleExistingEdge == existingSelectionEnd && !isEnd)) {
+          // If either edge exists and the opposite edge is being moved, then the
+          // opposite edge is moved to encompass the entire word at the new position.
+          if (position.offset < possibleExistingEdge.offset) {
+            targetPosition = wordBoundary.wordStart;
+          } else {
+            targetPosition = wordBoundary.wordEnd;
+          }
+        } else {
+          // Move edge to closest word boundary.
+          final int differenceA = (position.offset - wordBoundary.wordStart.offset).abs();
+          final int differenceB = (position.offset - wordBoundary.wordEnd.offset).abs();
+          targetPosition = differenceA < differenceB ? wordBoundary.wordStart : wordBoundary.wordEnd;
+        }
+      }
+    } else {
+      // The localPosition is not contained within the current rect. The adjustedPosition
+      // will either be at the end or beginning of the current rect.
+      if (existingSelectionStart != null && existingSelectionEnd != null) {
+        final TextPosition edgeToKeepInPlace = isEnd ? existingSelectionStart : existingSelectionEnd;
+        final bool shouldSwapEdgesWhenSelectionIsNormalized = isEnd ? position.offset < existingSelectionStart.offset : position.offset > existingSelectionEnd.offset;
+        final bool shouldSwapEdgesWhenSelectionNotNormalized = isEnd ? position.offset > existingSelectionStart.offset : position.offset < existingSelectionEnd.offset;
+
+        if (existingSelectionStart.offset < existingSelectionEnd.offset) {
+          if (shouldSwapEdgesWhenSelectionIsNormalized || position.offset == edgeToKeepInPlace.offset && _selectableContainsOriginWord!) {
+            final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(edgeToKeepInPlace);
+            assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+            _setSelectionPosition(isEnd ? localWordBoundary.wordEnd : localWordBoundary.wordStart, isEnd: !isEnd);
+          }
+        } else {
+          if (shouldSwapEdgesWhenSelectionNotNormalized || position.offset == edgeToKeepInPlace.offset && _selectableContainsOriginWord!) {
+            final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(edgeToKeepInPlace);
+            assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+            _setSelectionPosition(isEnd ? localWordBoundary.wordStart : localWordBoundary.wordEnd, isEnd: !isEnd);
+          }
+        }
+      }
+    }
+    targetPosition = _clampTextPosition(targetPosition ?? position);
+
+    _setSelectionPosition(targetPosition, isEnd: isEnd);
+    if (targetPosition.offset == range.end) {
```

Comment:
watch out for ellipsized case where the targetPosition may not be able to reach range.end. basically what to todo is describing below.

The return value should always be determine by `SelectionUtils.getResultBasedOnRect(_rect, localPosition);` until we have a way to know what text are ellipsized